### ISSUE TITLE
Apply_model does not yield same outputs as the trained model 

### ISSUE THREAD

**Issue by gourie** [dattalab/keypoint-moseq#143]
Hi,

First of all, big thank you and congrats for sharing this great tool with the community! 

I finished training a keypoint-moseq model and applied this pretrained model on new data, but ran into some consistency issues when qualitatively inspecting the syllables in a few videos which raised doubts on whether I am applying the inference (apply_model) function correctly.

To understand better what was going on, I decided to run a simpler experiment (see below) that yielded inconsistent results comparing the output of the apply_model function with the extract_results function. 
Can you help me understand why these results differ and how I can robustly apply a previously trained kp-moseq model to unseen data?

Details of my experiment: I compared the syllable results for a single video that was also used for training my model (in step 1)
1) right after training the model: load the most recent model checkpoint, then call extract_results() and save results to csv
        model, data, metadata, current_iter = kpms.load_checkpoint(PROJECT_DIR, MODEL_NAME)
        results = kpms.extract_results(model, metadata, PROJECT_DIR, MODEL_NAME)
        kpms.save_results_as_csv(results, PROJECT_DIR, MODEL_NAME)   
2) load the most recent model checkpoint, then run apply_model() on a single video of the training database (in a new inference_project_dir, re-using the same config.yml and the same DLC key points that were used for training the model):
        coordinates, confidences, bodyparts = kpms.load_keypoints(KEYPOINT_NEW_DATA_PATH, 'deeplabcut')
        data, metadata = kpms.format_data(coordinates, confidences, **config())
        model = kpms.load_checkpoint(MODEL_PATH, MODEL_NAME)[0]
        results = kpms.apply_model(model, data, metadata, INFERENCE_PROJECT_DIR, MODEL_NAME, save_results=True, results_path=os.path.join(INFERENCE_PROJECT_DIR, 'inference_results.csv'), **config())
        kpms.save_results_as_csv(results, INFERENCE_PROJECT_DIR, MODEL_NAME)   
3) compare the result csv files from 1) and 2) using diff function on Linux (or just less/cat both files and inspect them next to another)

Unfortunately, the syllables estimated for this single video were significantly different between methods 1) and 2), e.g., yielding significantly different histograms (i.e., certain syllables not being found a lot in 1) appeared at high counts in the outputs of 2)). 
I also found that repeating inference step 2) yielded small differences between the estimated syllables between two inference runs for v0.2.3 (the syllable histograms didn't differ much but there were many smaller differences in the estimated syllables across the time points) and tiny differences with v0.4.5 (across time points), using the same config and DLC key points as inputs.

Can you help me understand the above results and guide me how to robustly apply a previously trained kp-moseq model to new, unseen videos? 

Many thanks
Joeri

**Comment by calebweinreb** [comment]
Hi Joeri,

Keypoint-MoSeq learns syllables by fitting a Bayesian model. The syllable sequence that you get as output is (best case) a sample from the posterior distribution of that model. Even after the parameters are fixed (e.g. when applying a trained model), there is still a distribution over possible syllable sequences and therefore its expected that the precise sequence would differ from run to run. So that explains variation in results of the inference step. 

On top of this inherent stochasticity, there is also the issue of whether the sampling algorithm converges to the posterior within the number of iterations that you are using. It's possible that running the initial fitting and especially the apply_model steps with more iterations would improve convergence and could make the outputs of steps (1) and (2) more similar.

**Comment by jerrylchen** [comment]
I would like to reopen this conversation.   We are running into similar issues where we get different results applying the SAME  model repeatedly on the same trial data.  I found this to be very troubling from an interpretability standpoint.     How am I to believe that syllable  X is syllable X?

**Comment by calebweinreb** [comment]
Hi Jerry,

Good question- this variability is expected (to a certain degree) and explored in the paper (e.g. in Supp Fig. 3I). 

The core idea is that MoSeq is probabilistic, so it expresses a whole distribution of possible syllable sequences that are compatible with the data. In practice, we typically sample one such sequence and use it for downstream analysis, but a more rigorous approach would be to calculate these probabilities explicitly (which you can do with [this function](https://keypoint-moseq.readthedocs.io/en/latest/fitting.html#keypoint_moseq.fitting.estimate_syllable_marginals)). Here's what that looks like for the example explored in the paper (Supp Fig. 3I):

<img width="514" alt="Screenshot 2024-12-25 at 11 12 28 AM" src="https://github.com/user-attachments/assets/caa72f50-2c90-40ca-9105-0b032f8f25c6" />


You can see that the model has relatively high confidence at some points and is more 'confused' at others. The latter timepoints probably contain pose trajectories that are intermediate between several syllable prototypes. 

If the ambiguity in syllable assignments is especially concerning given the parameters of your project, one option would be to use the function linked above to only focus on those assignments that the model makes with high confidence (note that this is equivalent to applying the model many times with different random seeds and then counting how often each frame is assigned to its modal syllable).

**Comment by jerrylchen** [comment]
Thanks for the information.  I discussed this with Joeri and he seems to think that there is a bug in the code between how the syllables get indexed, or whatever lookup table is referenced   if one is using  extract_results  vs.   apply_model.  "syllable 1"  is not "syllable 1"  if you are not using the same function.  He advised that I should only use apply_model when I run the inference on trained data  vs.  new data   so that the syllable definition is consistent.

**Comment by calebweinreb** [comment]
That would be a big concern! The fact that syllables seem to change more than expected when running "apply_model" on data that was originally processed via "fit_model" is a known issue ([noted here](https://keypoint-moseq.readthedocs.io/en/latest/FAQs.html#detecting-existing-syllables-in-new-data) for example) so I think it's a good idea to uniformly process the whole dataset with "apply_model" instead of only applying this function to new data.

However, in my email correspondence with Joeri, there were a few observations that made me think this was purely a math/modeling issue as opposed to a code bug (for example, the problem went away when he excluded the full modeling step, and many timepoints remained the same, which is not what you would expect if the syllable labels were being fully permuted). I'd love to hear if Joeri has any new information on this issue!

**Comment by calebweinreb** [comment]
That makes total sense. One option would be to perform fitting and inference using just the "ar_only" phase of modeling (rather than including the subsequent "full model" phase). In the full model, syllables and pose are estimated simultaneously. In the ar-only model, syllables are fit directly to the empirical pose trajectory. The ar-only model tends to be more stable across runs and gives similar results when training a new model or fitting an existing model. The full model is especially useful when keypoint estimates are jittery/noisy or frequently missing. In your case, however, it might make better sense to deal with such noise (if there is any) using a more deterministic filtering approach.

**Comment by gourie** [comment]
I don't consider this to be a bug, yet I cannot explain why the syllables differ between the extract_results() and apply_model() functions. 
I confirm Caleb's statements that a model trained with AR_only did produce quasi-identical syllables from both functions and the apply_model function produced very similar syllables across different time points (num_iters argument).
I did verify at length the consistency and repeatability of the syllables when using only the apply_model() function and this does the job, hence I used the apply_model function all time in my experiments (as suggested by Caleb).
